lib/host/wasi/inode-win.cpp

```diff
@@ -180,51 +964,629 @@ WasiExpect<uint64_t> INode::getNativeHandler() const noexcept {
+  }
+
+  if (PathIsRelativeA(NewPath.c_str())) {
+    wchar_t HandleFullPathW[MAX_PATH];
+
+    // First get the paths of the handle
+    if (unlikely(GetFinalPathNameByHandleW(New.Handle, HandleFullPathW,
+                                           MAX_PATH, FILE_NAME_NORMALIZED)) ==
+        FALSE) {
+      return WasiUnexpect(fromWinError(GetLastError()));
+    }
+
+    if (!PathIsDirectoryW(HandleFullPathW)) {
+      return WasiUnexpect(__WASI_ERRNO_NOTDIR);
+    }
+
+    wchar_t NewPathW[MAX_PATH];
+
+    // Convert the path from char_t to wchar_t
+    mbstowcs(NewPathW, OldPath.c_str(), MAX_PATH);
+
+    // Append the paths together
+    HRESULT CombineResult =
+        PathCchCombine(NewFullPathW, MAX_PATH, HandleFullPathW, NewPathW);
+    switch (CombineResult) {
+    case S_OK:
+      break;
+    case E_INVALIDARG:
+      return WasiUnexpect(__WASI_ERRNO_INVAL);
+    case E_OUTOFMEMORY:
+      return WasiUnexpect(__WASI_ERRNO_OVERFLOW);
+    default:
+      return WasiUnexpect(__WASI_ERRNO_NAMETOOLONG);
+    }
+
+  } else {
+    mbstowcs(NewFullPathW, NewPath.c_str(), MAX_PATH);
+  }
+
+  // Create the hard link from the paths
+  if (unlikely(CreateHardLinkW(NewFullPathW, OldFullPathW, nullptr)) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  return {};
 }
 
-WasiExpect<INode> INode::pathOpen(std::string, __wasi_oflags_t,
-                                  __wasi_fdflags_t, uint8_t) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<INode> INode::pathOpen(std::string Path, __wasi_oflags_t OpenFlags,
+                                  __wasi_fdflags_t FdFlags,
+                                  uint8_t VFSFlags) const noexcept {
+  wchar_t FullPathW[MAX_PATH];
+
+  if (PathIsRelativeA(Path.c_str())) {
+    wchar_t HandleFullPathW[MAX_PATH];
+
+    // First get the paths of the handles
+    if (unlikely(GetFinalPathNameByHandleW(Handle, HandleFullPathW, MAX_PATH,
+                                           FILE_NAME_NORMALIZED)) == FALSE) {
+      return WasiUnexpect(fromWinError(GetLastError()));
+    }
+
+    if (!PathIsDirectoryW(HandleFullPathW)) {
+
+      return WasiUnexpect(__WASI_ERRNO_NOTDIR);
+    }
+
+    wchar_t PathW[MAX_PATH];
+
+    // Convert the path from char_t to wchar_t
+    mbstowcs(PathW, Path.c_str(), MAX_PATH);
+
+    // Append the paths together
+
+    HRESULT CombineResult =
+        PathCchCombine(FullPathW, MAX_PATH, HandleFullPathW, PathW);
+
+    switch (CombineResult) {
+    case S_OK:
+      break;
+    case E_INVALIDARG:
+      return WasiUnexpect(__WASI_ERRNO_INVAL);
+    case E_OUTOFMEMORY:
+      return WasiUnexpect(__WASI_ERRNO_OVERFLOW);
+    default:
+      return WasiUnexpect(__WASI_ERRNO_NAMETOOLONG);
+    }
+
+  } else {
+    mbstowcs(FullPathW, Path.c_str(), MAX_PATH);
+  }
+
+  DWORD AttributeFlags = attributeFlags(OpenFlags, FdFlags);
+  DWORD AccessFlags = accessFlags(FdFlags, VFSFlags);
+  // DWORD ShareFlags = shareFlags(OpenFlags);
```

Comment:
Do we need this?

Reply:
Related to this issue:
> Is it fine if multiple processes can access ,read and write potentially at the same time? Currently all functions allow external processes to read/write a file opened by WasmEdge. If yes then we should remove it else we may have to modify it is some way.

Reply:
Resolved in [here](https://github.com/WasmEdge/WasmEdge/pull/1826#discussion_r981949078)